README.md
# Consensus Port

![NuxtJS](https://img.shields.io/badge/Nuxt-black?style=for-the-badge&logo=nuxt.js&logoColor=white)
![TailwindCSS](https://img.shields.io/badge/tailwindcss-%2338B2AC.svg?style=for-the-badge&logo=tailwind-css&logoColor=white)
![Chart.js](https://img.shields.io/badge/chart.js-F5788D.svg?style=for-the-badge&logo=chart.js&logoColor=white)
![Docker](https://img.shields.io/badge/docker-%230db7ed.svg?style=for-the-badge&logo=docker&logoColor=white)
![NodeJS](https://img.shields.io/badge/node.js-6DA55F?style=for-the-badge&logo=node.js&logoColor=white)
![TypeScript](https://img.shields.io/badge/typescript-%23007ACC.svg?style=for-the-badge&logo=typescript&logoColor=white)

Accumulate Network block explorer for supply chain consensus.

## About 

Accumulate is a high performance blockchain for DeFi, integrating blockchains, protocols, oracles, and applications. This block explorer helps you understand and act upon transactions from supply chain oracle data on Accumulate, demonstrating how the network provides a bridge to the digital economy.

The default test supply chain oracle data for this explorer is on Accumulate testnet under the ADIs `acc://.../acme` and `acc://.../data`. If you'd like to use your own oracle ADIs to connect to your test data or supply chain, you can override the ADIs in `ADI_LIST` environment variable. This block explorer assumes a [standard schema](https://schema.org/) for supply chain oracle data, and you should structure your transactions according to this schema's required and extendable attributes. 

## Setup

Make sure to install the dependencies

```bash
npm install
```

## Development

Start the development server on http://localhost:3000

```bash
npm run dev
```

## Testing

We need to add test scripts compatible with Nuxt SSR and Docker (something like `@nuxt/test-utils` for Nuxt 3). Then we need to run those test scripts in `.github/workflows/pull-request.yaml` after the step for building the Docker image. In the longer term, we will also add better tooling for testing applications on local Kubernetes clusters that more closely resemble the production environment (more so than Kubernetes on Docker or K3s, using something like [Okteto](https://github.com/okteto/okteto))

## Releases

To create a new release according to the [Semantic Versioning](https://semver.org/) standard, we can use a CLI tool like [release-it](https://github.com/release-it/release-it). This helps us automate keeping our versioning synced across repositories and branches.

First run the following command:

```sh
npm run release
```

Then select the appropriate version increment and agree to commit the release to the git repository:

```sh
? Select increment (next version): patch (0.0.2) # major, minor, or patch
✔ npx auto-changelog -p

Changeset:
 M CHANGELOG.md
 M package-lock.json
 M package.json

? Commit (Release 0.0.2)? y
```

This sequence results in the following changes: 
- [CHANGELOG.md](CHANGELOG.md) is updated with all commits since the previous version
- [package.json](package.json), and [package-lock.json](package-lock.json) are updated with the new version
- File changes are committed to the repository 
- A new GitHub release is created with the new version as the tag

## Deployment

We use the following GitHub Actions workflows to test, build and push our images to [our organization](https://hub.docker.com/r/consensusnetworks/consensus-port) on Docker Hub: 
- Pull requests to `develop` and `master` branches trigger image builds (and eventually tests) 
- Pushes to `master` branch trigger image builds and pushes to the registry under the tags **latest** and **{your-commit-hash}**
- Releases from the `master` branch trigger image builds and pushes to the registry under the tags **latest**, **{your-commit-hash}**, and **{your-release-number}**

We can then manage our Kubernetes clusters and pod manifests in a separate repository to isolate environment configuration from application development. For reference, we included sample Kubernetes manifests that would be used in the separate repository to deploy this Nuxt app to a cloud provider managed cluster in the `kubernetes` directory. In addition to manifests, the separate repository would cover the following:
- Kubernetes cluster and managed infrastructure (like a DynamoDB) configuration
- Environment variables and secrets
- CI/CD pipelines for deploying integrated application resources to the cluster
- Helmfiles and Helm charts for organizing common application resources for the cluster

This repository setup also makes a serverless developer feel at home making cloud native applications (by hiding the ugly parts elsewhere).

## Contributing

Pull requests targeting the `develop` branch are welcome anytime. For major changes, please check out what's already being worked on [here](https://github.com/consensusnetworks/consensus-port/issues) and open a new issue before starting!

Please make sure to update tests as appropriate. We will be adding a style guide in the future.

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
Changelog All notable changes to this project will be documented in this file.

The format is based on Keep a Changelog, and this project adheres to Semantic Versioning.

composables/useAccumulate/index.ts
import { useNuxtApp } from "#app";
import {readonly, ref, onMounted} from "vue"



export const useAccumulate = (net: string): useAccumulate => {
    const { $instance: rpc } = useNuxtApp()

    const transactionHistory = ref<any>([])
    const transaction = ref<any>({})

    // defaults for the rpc request
    const defaults = {
      method: 'POST',
        headers: {'Content-Type': 'application/json'},
    };

    // each rpc request needs a unique id
    // a simple increment is enough, see below
    let id = 0

    // on mount, we want to get the transaction history
    onMounted(() => {
        console.log("on mount: get transaction history")


    })

    // get transaction by id
    async function getTransaction(txId: string): Promise<TransactionQueryResponse> {
        const tx = await rpc("https://testnet.accumulatenetwork.io/v2", {
            ...defaults,
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: id++,
                method: 'query-tx',
                params: {
                    txid: txId,
                },
            }),
        })

        if (!tx.ok) return rpcErrCodes[tx.status.toString()]

        const data = await tx.json()
        transaction.value = data.result

    }

    // get transaction history by address, start and count
     async function getTransactionHistory(prop: TransactionHistoryRequest): Promise<any> {
        try {
            const history = await rpc("https://testnet.accumulatenetwork.io/v2", {
                ...defaults,
                body: JSON.stringify({
                    jsonrpc: '2.0',
                    id: id++,
                    method: 'query-tx-history',
                    params: {
                        url: prop.url,
                        start: 0,
                        count: 100
                    },
                }),
            })

            if (!history.ok) return rpcErrCodes[history.status.toString()]

            const data = await history.json()
            transactionHistory.value = data.result.items
        } catch (err) {
            console.error(err)
        }
    }
    return {
        transaction: readonly(transaction),
        getTransaction,
        transactionHistory: readonly(transactionHistory),
        getTransactionHistory,
    }
 }

 export type useAccumulate = {
    transaction: any,
    transactionHistory: any,
    getTransactionHistory: ({ url, start, count }) => Promise<any>;
    getTransaction: (url: string, txid: string) => Promise<any>;
};

type AccumulateUrl = `acc://${string}/ACME`;

interface TransactionHistoryRequest {
  url: AccumulateUrl;
  start: number;
  count: number;
}

interface TransactionQueryResponse {
  jsonrpc: string;
  result: {
    type: string;
    data: {
      from: string;
      to: {
        url: string;
        amount: number;
        txid: string;
      }[];
    };
    sponsor: string;
    keyPage: {
      height: number;
    };
    txid: string;
    signer: {
      publicKey: string;
      nonce: number;
    };
    sig: string;
    status: {
      code: string;
    };
    syntheticTxids: string[];
  };
  id: number;
}

const rpcErrCodes = {
  "-32700": {
    code: -32700,
    message: "Parse error",
    meaning:
      "Invalid JSON was received by the server. An error occurred on the server while parsing the JSON text.",
  },
  "-32600": {
    code: -32600,
    message: "Invalid Request",
    meaning: "The JSON sent is not a valid Request object.",
  },
  "-32601": {
    code: -32601,
    message: "Method not found",
    meaning: "The method does not exist / is not available.",
  },
  "-32602": {
    code: -32602,
    message: "Invalid params",
    meaning: "Invalid method parameter(s).",
  },
  "-32603": {
    code: -32603,
    message: "Internal error",
    meaning: "Internal JSON-RPC error.",
  },
  "-32000": {
    code: -32000,
    message: "Server error",
    meaning: "Reserved for implementation-defined server-errors.",
  },
  "-32099": {
    code: -32099,
    message: "Server error",
    meaning: "Reserved for implementation-defined server-errors.",
  },
};


composables/useAccumulate.ts
import { useNuxtApp} from "#app";
import {readonly, ref} from "vue"

export const useAccumulate = (net: string): useAccumulate => {
    const { $instance: rpc } = useNuxtApp()

    const transactionHistory = ref<any>([])
    const transaction = ref<any>({})


    // defaults for the rpc request
    const defaults = {
      method: 'POST',
        headers: {'Content-Type': 'application/json'},
    };

    // each rpc request needs a unique id
    // a simple increment is enough, see below
    let id = 0

    let whichNet = net || "testnet"

    // get transaction by id
    async function getTransaction(txId: string): Promise<TransactionQueryResponse> {
        const tx = await rpc(`https://${whichNet}.accumulatenetwork.io/v2`, {
            ...defaults,
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: id++,
                method: 'query-tx',
                params: {
                    txid: txId,
                },
            }),
        })

        if (!tx.ok) return rpcErrCodes[tx.status.toString()]

        const data = await tx.json()

        transaction.value = data.result
    }

    // get transaction history by address, start and count
     async function getTransactionHistory(args: TransactionHistoryRequest): Promise<any> {
        console.log(args)
        try {
            const history = await rpc("https://testnet.accumulatenetwork.io/v2", {
                ...defaults,
                body: JSON.stringify({
                    jsonrpc: '2.0',
                    id: id++,
                    method: 'query-tx-history',
                    params: args
                }),
            })

            if (!history.ok) return rpcErrCodes[history.status.toString()]

            const data = await history.json()
            transactionHistory.value = data.result.items

        } catch (err) {
            console.error(err)
        }
    }
    return {
        transaction: readonly(transaction),
        getTransaction,
        transactionHistory: readonly(transactionHistory),
        getTransactionHistory,
    }
 }

 export type useAccumulate = {
    transaction: any,
    transactionHistory: any,
    getTransactionHistory: ({ url, start, count }) => Promise<any>;
    getTransaction: (url: string, txid: string) => Promise<any>;
};

type AccumulateUrl = `acc://${string}/ACME`;

interface TransactionHistoryRequest {
  url: AccumulateUrl;
  start: number;
  count: number;
}

interface TransactionQueryResponse {
  jsonrpc: string;
  result: {
    type: string;
    data: {
      from: string;
      to: {
        url: string;
        amount: number;
        txid: string;
      }[];
    };
    sponsor: string;
    keyPage: {
      height: number;
    };
    txid: string;
    signer: {
      publicKey: string;
      nonce: number;
    };
    sig: string;
    status: {
      code: string;
    };
    syntheticTxids: string[];
  };
  id: number;
}

const rpcErrCodes = {
  "-32700": {
    code: -32700,
    message: "Parse error",
    meaning:
      "Invalid JSON was received by the server. An error occurred on the server while parsing the JSON text.",
  },
  "-32600": {
    code: -32600,
    message: "Invalid Request",
    meaning: "The JSON sent is not a valid Request object.",
  },
  "-32601": {
    code: -32601,
    message: "Method not found",
    meaning: "The method does not exist / is not available.",
  },
  "-32602": {
    code: -32602,
    message: "Invalid params",
    meaning: "Invalid method parameter(s).",
  },
  "-32603": {
    code: -32603,
    message: "Internal error",
    meaning: "Internal JSON-RPC error.",
  },
  "-32000": {
    code: -32000,
    message: "Server error",
    meaning: "Reserved for implementation-defined server-errors.",
  },
  "-32099": {
    code: -32099,
    message: "Server error",
    meaning: "Reserved for implementation-defined server-errors.",
  },
};
